### ISSUE TITLE
@ConditionalOnBean in a @Configuration @Bean methods fails when depending on another bean set up by @Bean

### ISSUE THREAD

**Issue by froh42** [spring-projects/spring-framework#26954]
# Problem

I do have to separate classes marked as @Configuration, both set up beans with bean factory methods annotated with @Bean (please correct me if I have the wording off, I'm not an expert).

The first:

        @Configuration
        public class KidneyConfig {
        
             @Bean
            KidneyBean makeKidneyBean() { return new KidneyBean() }
        }

The dependent one:

        public class ChiliConfig {
             @ConditionalOnBean(KidneyBean.class)
             @Bean
             Chili makeChili(KidneyBean kidneyBean) { return new Chili(kidenyBean); ]
        }

The problem: Building everything as an application into a jar file (spring boot application, but I don't think that plays a role here) I can sometimes start the application, sometimes it will fail claiming that the condition failed.

# Analysis 

(I learned Spring internals on the go analyzing this bug, so please bear with inaccuracies):

When an ApplicationContext is built it will call refresh, and from there do a classpath scan to detect beans. The classpath scan is order dependent, so the bean definitions are in different order in the ApplicationContext.

Later in refresh() Spring will happen to call the ConfigurationClassParser which will scan @Configuration classes for @Bean methods.
If it finds one it will use ConfigurationClass to locate the Methods annontated with @Bean

ConfigurationClass in turn will evaluate the Conditionals right away in this moment, so if my KidneyBean was already scanned by ConfigurationClass then everything will start up fine.

But if the Chili is scanned earlier than the KidneyBean, then the @ConditionalOnBean will fail - although just a moment later the ConfigurationClassParser will load the KidneyBean.

I'm not qualified to suggest how to fix this.

**Comment by froh42** [comment]
Sorry, for the changes in the title, had a few typos in the comment after a long debugging session.

**Comment by froh42** [comment]
I have pushed  a small demo showing this problem to https://github.com/froh42/spring-issue-demo-26964.

Its readme tries to explain it again, now that I'm not sleepy.

The demo application has both a ChiliConfig and a KidneyConfig in the situation described above. Starting it, the start will fail because ChiliConfig and its ConditionalOnBean are evaluated before KidneyConfig.

Renaming the KidneyConfig to AAAKidneyConfig will result in the KindneyConfig being loaded before ChiliConfig and the application will run.

**Comment by sbrannen** [comment]
Hi @froh42,

I am not quite sure why you are using `@ConditionalOnBean` in your sample application.

If you remove the use of `@ConditionalOnBean` from your `makeChiliBean()` method, your application starts up correctly.

Please note that declaring the dependency on a `KidneyBean` via a method parameter in `makeChiliBean()` instructs Spring that `makeChiliBean()` cannot be invoked without a `KidneyBean` present in the `ApplicationContext`.

Does that suit your needs?

If not, please explain what you are trying to achieve by using `@ConditionalOnBean`.

**Comment by froh42** [comment]
Hi @sbrannen , yes in the sample application I don't need the @ConditionalOnBean at all - and in my real application I have worked around the problem by switching away @ConditionalOnBean to @ConditionalOnProperty (using a property to control that part of the application is a valid solution for me).

The sample application includes the @ConditionalOnBean, because it is simply required to trigger the problem. Yes, when I remove it, it starts up correctly.

The sample is just the most reduced version of some code I could come up with to showcase the problem.


I did find this by finding out (and working around) the issue that the applicaton my team is developing did not start up in certain cases. I can't publish that one here, unfortunately.

What I can do: In the demo application, I can document the place where to set a breakpoint and condition for the breakpoint, so it is easy to debug to exactly the place where I think the problem occurs. I'll do that right away and add another comment, then.

**Comment by froh42** [comment]
Hi @sbrannen again, I updated the https://github.com/github/renaming#readme on the demo app with what I found out about where the problem exactly happens. (Stack traces, screenshots and my interpretation of things).

**Comment by ttddyy** [comment]
@froh42 I had bit by similar case before. 
You should only use `@ConditionalOnBean` in auto-configuration classes as it comes from `autoconfigure` module in Spring Boot.

`@ConditionalOnBean` is dependent on the order of processing `@Configuration` classes.

The javadoc clearly states the order dependency and recommends the usage in auto configurations.

> The condition can only match the bean definitions that have been processed by the application context so far and, as such, it is strongly recommended to use this condition on auto-configuration classes only. If a candidate bean may be created by another auto-configuration, make sure that the one using this condition runs after.

The auto-configuration classes are guaranteed to be processed after user configurations. Also, it has a mechanism to force relative orders within auto configurations by `@AutoConfigure[Before|After]`.

Without a mechanism to order the configuration classes, the order of processing the `@Configuration` class that has `@ConditionalOnBean` is not deterministic. Therefore, you should only use it within auto configuration as it is the intended usage of the condition.

**Comment by froh42** [comment]
@ttddyy yes, I do aggree. Of course I didn't re-read the documentation before applying the `@ConditionalOnBean`

That's why a good mitigation for the problem might just be printing just a warning if you DO use `@ConditionalOnBean` in this manner - this would have shortened debugging time. I do think a `log.warning` could be called in that case, that just says that.

People like me - stumbiling into this trap - would then find at least a string to google for and find the documentation.

**Comment by sbrannen** [comment]
> I have worked around the problem by switching away @ConditionalOnBean to @ConditionalOnProperty (using a property to control that part of the application is a valid solution for me).

Have you considered using bean definition profiles (`@Profile(...)`) instead of `@ConditionalOnProperty`?

I think profiles might be a better fit for your use case.

As @ttddyy pointed out, `@ConditionalOnBean` is not intended to be used the way you were using it.

@ttddyy, thanks for providing your detailed answer!

> Do you think, detecting this situation and adding a warning log message (or event debug log message, when you start with --debug argument) might be a good call? _I_ learned my lesson, but other might experience this as well.

Since `@ConditionalOnBean` is a Spring Boot feature, I'll let the Spring Boot team decide if they wish to implement some form of warning for using `@ConditionalOnBean` in a `@Configuration` class that is not participating in auto-configuration.
